### PR TITLE
Keras model building

### PR DESCRIPTION
**Pull request by delhomer**
Implement dedicated classes to build feature detection and semantic segmentation architectures. It is crucial to notice that the produced classes correspond to **architectures** instead of **models** for now. We will use the Keras API to easily create models, by considering `Network.X` and `Network.Y` : 
```
net = FeatureDetectionNetworl(image_size=224, nb_channels=3, nb_labels=65)
model = keras.models.Model(net.X, net.Y)
model.compile(...)
model.summary() # Gives the whole architecture of the network, built during the instanciation of `net`
```

In this PR, a first example of state-of-the-art model architecture reusing is also provided with `VGG16` network (see *Simonyan, 2015*). For more details, see `architecture` attribute of class `FeatureDetectionNetwork` and `SemanticSegmentation`.

Some unit tests come with these new features.

### CODE REVIEW

File: tests/test_feature_detection.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+"""Unit test related to feature detection model instanciation
```
Comment (garaud) [inline tests/test_feature_detection.py]: I would have written "Unit test related to the feature detection model"

I don't think that the word `instanciation` is correct.

File: tests/test_feature_detection.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+"""
+
+from keras.models import Model
+
+from deeposlandia.keras_feature_detection import FeatureDetectionNetwork
+
+def test_simple_network_architecture():
+    """Test the instanciation of a simple feature detection network
```
Comment (garaud) [inline tests/test_feature_detection.py]: Could you remove the word `instanciation`?

File: tests/test_feature_detection.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    m = Model(net.X, net.Y)
+    input_shape = m.input_shape
+    output_shape = m.output_shape
+    assert len(input_shape) == 4
+    assert input_shape[1:] == (IMAGE_SIZE, IMAGE_SIZE, NB_CHANNELS)
+    assert len(output_shape) == 2
+    assert output_shape[1] == NB_LABELS
+
+def test_vgg16_network_architecture():
+    """Test the instanciation of a VGG16-inspired feature detection network
```
Comment (garaud) [inline tests/test_feature_detection.py]: Could you remove the word "instanciation"?

File: tests/test_semantic_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+"""Unit test related to semantic segmentation model instanciation
```
Comment (garaud) [inline tests/test_semantic_segmentation.py]: Unit test related to the semantic segmentation model.

File: tests/test_semantic_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+"""
+
+from keras.models import Model
+
+from deeposlandia.keras_semantic_segmentation import SemanticSegmentationNetwork
+
+def test_network_instanciation():
+    """Test the instanciation of a simple feature detection network
```
Comment (garaud) [inline tests/test_semantic_segmentation.py]: Could you remove the word "instanciation"?

File: tests/test_feature_detection.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                                  nb_channels=NB_CHANNELS, nb_labels=NB_LABELS)
+    m = Model(net.X, net.Y)
+    input_shape = m.input_shape
+    output_shape = m.output_shape
+    assert len(input_shape) == 4
+    assert input_shape[1:] == (IMAGE_SIZE, IMAGE_SIZE, NB_CHANNELS)
+    assert len(output_shape) == 2
+    assert output_shape[1] == NB_LABELS
+
+def test_vgg16_network_architecture():
```
Comment (garaud) [inline tests/test_feature_detection.py:25]: It seems that this test is quite long (several seconds). What do you think? Do you know why?

File: deeposlandia/network.py [hunk 7]
```diff
@@ -5,20 +5,32 @@
 
 class ConvolutionalNeuralNetwork:
     """Convolutional neural network design
+
+    Attributes
+    ----------
+    _network_name : str
```
Comment (garaud) [inline deeposlandia/network.py]: It's quite strange for me to have some "protected" attributes in the docstring. The most of the time, you write `self._attr` when you don't want to let it be "public", so you don't write a docstring for it.

I prefer to have "public" and explicit attributes which define you typed object. Is it possible to replace the `self._attr` by `self.attr`?

File: deeposlandia/keras_feature_detection.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        Number of classes in the dataset glossary
+    X : tensor
+        (batch_size, image_size, image_size, nb_channels)-shaped input tensor; input image data
+    Y : tensor
+        (None, nb_classes)-shaped output tensor, built as the output of the last network layer
+    """
+
+    def __init__(self, network_name="mapillary", image_size=512, nb_channels=3,
+                 nb_labels=65, learning_rate=1e-4, architecture="simple"):
+        ConvolutionalNeuralNetwork.__init__(self, network_name, image_size,
```
Comment (garaud) [inline deeposlandia/keras_feature_detection.py]: I think you should use the `super()` function for the inheritance.

File: deeposlandia/network.py [hunk 9]
```diff
@@ -5,20 +5,32 @@
...
         self._network_name = network_name
         self._image_size = image_size
         self._nb_channels = nb_channels
         self._nb_labels = nb_labels
-        self._X = K.layers.Input(shape=(image_size, image_size, nb_channels))
+        self.X = K.layers.Input(shape=(image_size, image_size, nb_channels), name="input")
 
     def convolution(self, x, nb_filters, kernel_size, strides=1, padding="same",
-                    activation='relu', batch_norm=True):
+                    activation='relu', batch_norm=True, name=""):
```
Comment (garaud) [inline deeposlandia/network.py]: The Keras layer, by default, defines a layer name according to its type, e.g. `Conv2D`, with an incremented integer such as `conv2d_2`, `conv2d_3`, and so on.

I think we should set it `None` by default. Then, Keras defines its own. And if we want to specify it, we can.

File: deeposlandia/network.py [hunk 10]
```diff
@@ -137,17 +155,35 @@ def dense(self, x, depth, dropout_rate=1.0, activation='relu', batch_norm=True):
...
         if batch_norm:
-            x = K.layers.BatchNormalization()(x)
-        x = K.layers.Activation(activation)(x)
-        x = K.layers.Dropout(dropout_rate)(x)
+            x = K.layers.BatchNormalization(name=name+'_bn')(x)
+        x = K.layers.Activation(activation, name=name+'_activation')(x)
+        x = K.layers.Dropout(dropout_rate, name=name+'_dropout')(x)
         return x
 
+    def flatten(self, x, name=""):
```
Comment (garaud) [inline deeposlandia/network.py]: Same comment about the `name` parameter.

**Review by garaud — CHANGES_REQUESTED** [review]
The `Model(nn.X, nn.Y)` is a very good idea. Some comment with minor changes.

File: deeposlandia/network.py (hunk 7 above)
Comment (delhomer) [reply deeposlandia/network.py]: Yes, these stuffs are remainders of a quite clumsy C++-convention adaptation. Useless in Python, I remove them.

File: deeposlandia/network.py (hunk 9 above)
Comment (delhomer) [reply deeposlandia/network.py]: Yes indeed, I saw the Keras default behavior for layer naming. `None` is a better idea than empty string, for sure. I change this.

File: deeposlandia/network.py (hunk 10 above)
Comment (delhomer) [reply deeposlandia/network.py]: Seen.

File: tests/test_feature_detection.py (hunk 1 above)
Comment (delhomer) [reply tests/test_feature_detection.py]: `instanciation` word has been removed.

File: tests/test_feature_detection.py (hunk 2 above)
Comment (delhomer) [reply tests/test_feature_detection.py]: Done.

File: tests/test_feature_detection.py (hunk 3 above)
Comment (delhomer) [reply tests/test_feature_detection.py]: Done.

File: tests/test_semantic_segmentation.py (hunk 4 above)
Comment (delhomer) [reply tests/test_semantic_segmentation.py]: Done.

File: tests/test_semantic_segmentation.py (hunk 5 above)
Comment (delhomer) [reply tests/test_semantic_segmentation.py]: Done.

File: deeposlandia/keras_feature_detection.py (hunk 8 above)
Comment (delhomer) [reply deeposlandia/keras_feature_detection.py]: Changed to `super().__init__(network_name, image_size, ...)`.
This format is only `Python-3.x` compliant, however I guess the code is still Python3-oriented. We mention Python3 in the `setup.py`, I do not if we have to specify again elsewhere.

File: tests/test_feature_detection.py (hunk 6 above)
Comment (delhomer) [reply tests/test_feature_detection.py:25]: Hum, I'm quite surprised, it is less than one second on my dev machine.

However, the fact it can be long is not surprising: we are instantiating a deep neural network with around 130 millions parameters here. Some time is needed to create and initialize all the weights, I guess.

**Comment by garaud** [comment]
We mention Python 3 in the README.md file
Raphael Delhome writes:

> delhomer commented on this pull request.
>
>
>
>> +        Name of the network
> +    _image_size : integer
> +        Input image size (height and width are equal)
> +    _nb_channels : integer
> +        Number of input image channels (1 for greyscaled images, 3 for RGB images)
> +    _nb_labels : integer
> +        Number of classes in the dataset glossary
> +    X : tensor
> +        (batch_size, image_size, image_size, nb_channels)-shaped input tensor; input image data
> +    Y : tensor
> +        (None, nb_classes)-shaped output tensor, built as the output of the last network layer
> +    """
> +
> +    def __init__(self, network_name="mapillary", image_size=512, nb_channels=3,
> +                 nb_labels=65, learning_rate=1e-4, architecture="simple"):
> +        ConvolutionalNeuralNetwork.__init__(self, network_name, image_size,
>
> Changed to `super().__init__(network_name, image_size, ...)`.
> This format is only `Python-3.x` compliant, however I guess the code is still Python3-oriented. We mention Python3 in the `setup.py`, I do not if we have to specify again elsewhere.


-- 
Damien Garaud
Data Scientist – Developer, Oslandia
http://oslandia.com

**Comment by delhomer** [comment]
I pushed another commit so as to fix the case when layer names are equal to `None`.

Now, if a hand-made layer has a `block_name` equal to None, all the layers that are contained into it take `None` as name (*e.g.* for `convolution()`, that contains a `Conv2D`, a `BatchNormalization()` and a `Activation()` native layers).
